File: app/src/main/java/com/example/btlapp/MockTestActivity.java
package com.example.btlapp;

import android.os.Bundle;
import android.os.CountDownTimer;
import android.view.View;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MockTestActivity extends AppCompatActivity {

    private List<Question> questions;
    private int currentIndex = 0;
    private Map<Integer, Integer> userAnswers = new HashMap<>(); // Question Index -> Selected Option

    private TextView tvTimer;
    private TextView tvCount;
    private TextView tvContent;
    private RadioGroup rgOptions;
    private RadioButton rbA;
    private RadioButton rbB;
    private RadioButton rbC;
    private RadioButton rbD;
    private Button btnPrev;
    private Button btnNext;
    private Button btnSubmit;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_mock_test);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        initViews();
        loadMockQuestions();
        startTimer();
        displayQuestion();

        btnNext.setOnClickListener(v -> {
            saveAnswer();
            if (currentIndex < questions.size() - 1) {
                currentIndex++;
                displayQuestion();
            }
        });

        btnPrev.setOnClickListener(v -> {
            saveAnswer();
            if (currentIndex > 0) {
                currentIndex--;
                displayQuestion();
            }
        });

        btnSubmit.setOnClickListener(v -> {
            saveAnswer();
            showSubmitDialog();
        });
    }

    private void initViews() {
        tvTimer = findViewById(R.id.tvTimer);
        tvCount = findViewById(R.id.tvMockQuestionCount);
        tvContent = findViewById(R.id.tvMockQuestionContent);
        rgOptions = findViewById(R.id.rgMockOptions);
        rbA = findViewById(R.id.rbMockOptionA);
        rbB = findViewById(R.id.rbMockOptionB);
        rbC = findViewById(R.id.rbMockOptionC);
        rbD = findViewById(R.id.rbMockOptionD);
        btnPrev = findViewById(R.id.btnMockPrevious);
        btnNext = findViewById(R.id.btnMockNext);
        btnSubmit = findViewById(R.id.btnSubmit);
    }

    private void loadMockQuestions() {
        String type = getIntent().getStringExtra("LICENSE_TYPE");
        if (type == null) type = "MOTORBIKE";
        
        List<Question> allQuestions;
        if (type.equals("MOTORBIKE")) {
            allQuestions = QuestionRepository.getMotorbikeQuestions();
        } else {
            allQuestions = QuestionRepository.getCarQuestions();
        }
        
        Collections.shuffle(allQuestions);
        questions = allQuestions.subList(0, Math.min(20, allQuestions.size()));
    }

    private void startTimer() {
        new CountDownTimer(20 * 60 * 1000, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                long minutes = millisUntilFinished / 1000 / 60;
                long seconds = (millisUntilFinished / 1000) % 60;
                tvTimer.setText(String.format("%02d:%02d", minutes, seconds));
            }

            @Override
            public void onFinish() {
                tvTimer.setText("00:00");
                calculateResult();
            }
        }.start();
    }

    private void displayQuestion() {
        Question q = questions.get(currentIndex);
        tvCount.setText("Câu " + (currentIndex + 1) + "/" + questions.size());
        tvContent.setText(q.getContent());
        rbA.setText(q.getOptionA());
        rbB.setText(q.getOptionB());
        
        if (q.getOptionC() != null) {
            rbC.setVisibility(View.VISIBLE);
            rbC.setText(q.getOptionC());
        } else {
            rbC.setVisibility(View.GONE);
        }
        
        if (q.getOptionD() != null) {
            rbD.setVisibility(View.VISIBLE);
            rbD.setText(q.getOptionD());
        } else {
            rbD.setVisibility(View.GONE);
        }

        rgOptions.clearCheck();
        Integer answer = userAnswers.get(currentIndex);
        if (answer != null) {
            if (answer == 1) rbA.setChecked(true);
            else if (answer == 2) rbB.setChecked(true);
            else if (answer == 3) rbC.setChecked(true);
            else if (answer == 4) rbD.setChecked(true);
        }
    }

    private void saveAnswer() {
        int selectedId = rgOptions.getCheckedRadioButtonId();
        if (selectedId != -1) {
            int answer = 0;
            if (selectedId == R.id.rbMockOptionA) answer = 1;
            else if (selectedId == R.id.rbMockOptionB) answer = 2;
            else if (selectedId == R.id.rbMockOptionC) answer = 3;
            else if (selectedId == R.id.rbMockOptionD) answer = 4;
            
            userAnswers.put(currentIndex, answer);
        }
    }

    private void showSubmitDialog() {
        new AlertDialog.Builder(this)
            .setTitle("Nộp bài")
            .setMessage("Bạn có chắc chắn muốn nộp bài?")
            .setPositiveButton("Có", (dialog, which) -> calculateResult())
            .setNegativeButton("Không", null)
            .show();
    }

    private void calculateResult() {
        int correctCount = 0;
        boolean failedCritical = false;

        for (int i = 0; i < questions.size(); i++) {
            Question question = questions.get(i);
            Integer userAnswer = userAnswers.get(i);
            if (userAnswer != null && userAnswer == question.getCorrectAnswer()) {
                correctCount++;
            } else if (question.isCritical()) {
                if (userAnswer != null) {
                    failedCritical = true;
                }
            }
        }

        String resultMessage;
        if (failedCritical) {
            resultMessage = "Bạn đã TRƯỢT do trả lời sai câu hỏi điểm liệt.";
        } else if (correctCount >= 18) {
            resultMessage = "Chúc mừng! Bạn đã ĐẠT (" + correctCount + "/" + questions.size() + ").";
        } else {
            resultMessage = "Bạn không đạt (" + correctCount + "/" + questions.size() + ").";
        }

        new AlertDialog.Builder(this)
            .setTitle("Kết quả")
            .setMessage(resultMessage)
            .setPositiveButton("Đóng", (dialog, which) -> finish())
            .setCancelable(false)
            .show();
    }
}

File: app/src/main/java/com/example/btlapp/MockTestActivity.kt
package com.example.btlapp

import android.os.Bundle
import android.os.CountDownTimer
import android.view.View
import android.widget.Button
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.TextView
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

class MockTestActivity : AppCompatActivity() {

    private lateinit var questions: List<Question>
    private var currentIndex = 0
    private val userAnswers = mutableMapOf<Int, Int>() // Question Index -> Selected Option

    private lateinit var tvTimer: TextView
    private lateinit var tvCount: TextView
    private lateinit var tvContent: TextView
    private lateinit var rgOptions: RadioGroup
    private lateinit var rbA: RadioButton
    private lateinit var rbB: RadioButton
    private lateinit var rbC: RadioButton
    private lateinit var rbD: RadioButton
    private lateinit var btnPrev: Button
    private lateinit var btnNext: Button
    private lateinit var btnSubmit: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_mock_test)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        initViews()
        loadMockQuestions()
        startTimer()
        displayQuestion()

        btnNext.setOnClickListener {
            saveAnswer()
            if (currentIndex < questions.size - 1) {
                currentIndex++
                displayQuestion()
            }
        }

        btnPrev.setOnClickListener {
            saveAnswer()
            if (currentIndex > 0) {
                currentIndex--
                displayQuestion()
            }
        }

        btnSubmit.setOnClickListener {
            saveAnswer()
            showSubmitDialog()
        }
    }

    private fun initViews() {
        tvTimer = findViewById(R.id.tvTimer)
        tvCount = findViewById(R.id.tvMockQuestionCount)
        tvContent = findViewById(R.id.tvMockQuestionContent)
        rgOptions = findViewById(R.id.rgMockOptions)
        rbA = findViewById(R.id.rbMockOptionA)
        rbB = findViewById(R.id.rbMockOptionB)
        rbC = findViewById(R.id.rbMockOptionC)
        rbD = findViewById(R.id.rbMockOptionD)
        btnPrev = findViewById(R.id.btnMockPrevious)
        btnNext = findViewById(R.id.btnMockNext)
        btnSubmit = findViewById(R.id.btnSubmit)
    }

    private fun loadMockQuestions() {
        val type = intent.getStringExtra("LICENSE_TYPE") ?: "MOTORBIKE"
        val allQuestions = if (type == "MOTORBIKE") {
            QuestionRepository.getMotorbikeQuestions()
        } else {
            QuestionRepository.getCarQuestions()
        }
        // In a real app, shuffle and take 25. Here we take all for demo.
        questions = allQuestions.shuffled().take(20)
    }

    private fun startTimer() {
        object : CountDownTimer(20 * 60 * 1000, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                val minutes = millisUntilFinished / 1000 / 60
                val seconds = (millisUntilFinished / 1000) % 60
                tvTimer.text = String.format("%02d:%02d", minutes, seconds)
            }

            override fun onFinish() {
                tvTimer.text = "00:00"
                calculateResult()
            }
        }.start()
    }

    private fun displayQuestion() {
        val q = questions[currentIndex]
        tvCount.text = "Câu ${currentIndex + 1}/${questions.size}"
        tvContent.text = q.content
        rbA.text = q.optionA
        rbB.text = q.optionB
        
        rbC.visibility = if (q.optionC != null) View.VISIBLE else View.GONE
        rbC.text = q.optionC ?: ""
        
        rbD.visibility = if (q.optionD != null) View.VISIBLE else View.GONE
        rbD.text = q.optionD ?: ""

        rgOptions.clearCheck()
        userAnswers[currentIndex]?.let {
            when (it) {
                1 -> rbA.isChecked = true
                2 -> rbB.isChecked = true
                3 -> rbC.isChecked = true
                4 -> rbD.isChecked = true
            }
        }
    }

    private fun saveAnswer() {
        val selectedId = rgOptions.checkedRadioButtonId
        if (selectedId != -1) {
            val answer = when (selectedId) {
                R.id.rbMockOptionA -> 1
                R.id.rbMockOptionB -> 2
                R.id.rbMockOptionC -> 3
                R.id.rbMockOptionD -> 4
                else -> 0
            }
            userAnswers[currentIndex] = answer
        }
    }

    private fun showSubmitDialog() {
        AlertDialog.Builder(this)
            .setTitle("Nộp bài")
            .setMessage("Bạn có chắc chắn muốn nộp bài?")
            .setPositiveButton("Có") { _, _ -> calculateResult() }
            .setNegativeButton("Không", null)
            .show()
    }

    private fun calculateResult() {
        var correctCount = 0
        var failedCritical = false

        questions.forEachIndexed { index, question ->
            val userAnswer = userAnswers[index]
            if (userAnswer == question.correctAnswer) {
                correctCount++
            } else if (question.isCritical) {
                failedCritical = true
            }
        }

        val resultMessage = if (failedCritical) {
            "Bạn đã TRƯỢT do trả lời sai câu hỏi điểm liệt."
        } else if (correctCount >= 18) {
            "Chúc mừng! Bạn đã ĐẠT ($correctCount/${questions.size})."
        } else {
            "Bạn không đạt ($correctCount/${questions.size})."
        }

        AlertDialog.Builder(this)
            .setTitle("Kết quả")
            .setMessage(resultMessage)
            .setPositiveButton("Đóng") { _, _ -> finish() }
            .setCancelable(false)
            .show()
    }
}
